crates/ty_server/tests/e2e/pull_diagnostics.rs

```diff
@@ -27,3 +30,150 @@ def foo() -> str:
+        .with_file(file_c, file_c_content_v1)?
+        .with_file(file_d, file_d_content_v1)?
+        .enable_pull_diagnostics(true)
+        .build()?
+        .wait_until_workspaces_are_initialized()?;
+
+    server.open_text_document(file_a, &file_a_content, 1);
+
+    // First request with no previous result IDs
+    let first_response = server.workspace_diagnostic_request(None)?;
+    insta::assert_debug_snapshot!("workspace_diagnostic_initial_state", first_response);
+
+    // Extract result IDs from the first response
+    let previous_result_ids = match first_response {
+        WorkspaceDiagnosticReportResult::Report(report) => {
+            report.items.into_iter().filter_map(|item| {
+                match item {
+                    WorkspaceDocumentDiagnosticReport::Full(full_report) => {
+                        let result_id = full_report.full_document_diagnostic_report.result_id?;
+                        Some(PreviousResultId {
+                            uri: full_report.uri,
+                            value: result_id,
+                        })
+                    }
+                    WorkspaceDocumentDiagnosticReport::Unchanged(_) => {
+                        // Shouldn't happen in the first request
+                        None
+                    }
+                }
+            })
+        }
+        WorkspaceDiagnosticReportResult::Partial(_) => {
+            panic!("The first response must be a full report");
+        }
+    }
+    .collect();
+
+    // Make changes to files B, C, and D (leave A unchanged)
+    // Need to open files before changing them
+    server.open_text_document(file_b, &file_b_content_v1, 1);
+    server.open_text_document(file_c, &file_c_content_v1, 1);
+    server.open_text_document(file_d, &file_d_content_v1, 1);
+
+    // File B: Add a new error
+    server.change_text_document(
+        file_b,
+        vec![lsp_types::TextDocumentContentChangeEvent {
+            range: None,
+            range_length: None,
+            text: file_b_content_v2.to_string(),
+        }],
+        2,
+    );
+
+    // File C: Fix the error
+    server.change_text_document(
+        file_c,
+        vec![lsp_types::TextDocumentContentChangeEvent {
+            range: None,
+            range_length: None,
+            text: file_c_content_v2.to_string(),
+        }],
+        2,
+    );
+
+    // File D: Change the error
+    server.change_text_document(
+        file_d,
+        vec![lsp_types::TextDocumentContentChangeEvent {
+            range: None,
+            range_length: None,
+            text: file_d_content_v2.to_string(),
+        }],
+        2,
+    );
+
+    // Second request with previous result IDs
+    // Expected results:
+    // - File A: Unchanged report (diagnostic hasn't changed)
+    // - File B: Full report (new diagnostic appeared)
+    // - File C: Full report with empty diagnostics (diagnostic was removed)
+    // - File D: Full report (diagnostic content changed)
```

Comment:
Should we add a test case where the diagnostic content didn't change but the range did? Like, new content got added to the file but the new content doesn't produce the diagnostics so the existing diagnostic just got moved further down.